sql/lease.go

```diff
@@ -0,0 +1,546 @@
+}
+
+// LeaseState holds the state for a lease. Exported only for testing.
+type LeaseState struct {
+	TableDescriptor
+	expiration int64
+	refcount   int
+}
+
+func (s *LeaseState) String() string {
+	return fmt.Sprintf("%d:%d", s.Version, s.expiration)
+}
+
+// Expiration returns the expiration time of the lease.
+func (s *LeaseState) Expiration() time.Time {
+	return nanosToTime(s.expiration)
+}
+
+// Refcount returns the reference count of the lease.
+func (s *LeaseState) Refcount() int {
+	return s.refcount
+}
+
+// LeaseStore implements the operations for acquiring and releasing leases and
+// publishing a new version of a descriptor. Exported only for testing.
+type LeaseStore struct {
+	db     client.DB
+	clock  *hlc.Clock
+	nodeID uint32
+}
+
+// jitteredLeaseDuration returns a randomly jittered duration from the interval
+// [0.75 * leaseDuration, 1.25 * leaseDuration].
+func (s LeaseStore) jitteredLeaseDuration() int64 {
+	return int64(float64(leaseDuration) * (0.75 + 0.5*rand.Float64()))
+}
+
+// Acquire a lease on the most recent version of a table descriptor.
+func (s LeaseStore) Acquire(tableID ID, minVersion uint32) (*LeaseState, error) {
+	lease := &LeaseState{}
+	lease.expiration = s.clock.Now().WallTime + s.jitteredLeaseDuration()
+	descKey := MakeDescMetadataKey(tableID)
+
+	err := s.db.Txn(func(txn *client.Txn) error {
+		if err := txn.GetProto(descKey, &lease.TableDescriptor); err != nil {
+			return err
+		}
+		if err := lease.Validate(); err != nil {
+			return err
+		}
+		if lease.Version < minVersion {
+			return fmt.Errorf("version %d of table %d does not exist yet", minVersion, tableID)
+		}
+		// Add an entry to the lease table.
+		leaseKey := makeLeaseKey(s.nodeID, lease.ID, lease.Version, lease.expiration)
+		if log.V(3) {
+			log.Infof("acquire: %s", prettyKey(leaseKey, 0))
+		}
+		b := &client.Batch{}
+		b.Put(leaseKey, nil)
+		txn.SetSystemDBTrigger()
+		return txn.CommitInBatch(b)
+	})
+	return lease, err
+}
+
+// Release a previously acquired table descriptor lease.
+func (s LeaseStore) Release(lease *LeaseState) error {
+	leaseKey := makeLeaseKey(s.nodeID, lease.ID, lease.Version, lease.expiration)
+	if log.V(3) {
+		log.Infof("release: %s", prettyKey(leaseKey, 0))
+	}
+	return s.db.Txn(func(txn *client.Txn) error {
+		b := &client.Batch{}
+		b.Del(leaseKey)
+		txn.SetSystemDBTrigger()
+		return txn.CommitInBatch(b)
+	})
+}
+
+// Publish a new version of a table descriptor. The update closure may be
+// called multiple times if retries occur: make sure it does not have side
+// effects.
+func (s LeaseStore) Publish(tableID ID, update func(*TableDescriptor) error) error {
+	desc := &TableDescriptor{}
+	descKey := MakeDescMetadataKey(tableID)
+
+	retryOpts := retry.Options{
+		InitialBackoff: 20 * time.Millisecond,
+		MaxBackoff:     2 * time.Second,
+		Multiplier:     2,
+	}
+
+	for r := retry.Start(retryOpts); r.Next(); {
+		// Get the current version of the table descriptor non-transactionally.
```

Comment:
this is just a single read, so it (almost) doesn't matter that you're not doing it in a transaction. Did you want to read inconsistently?